src/PenSkin.js

```diff
@@ -516,66 +275,36 @@ class PenSkin extends Skin {
         const attachments = [
             {
                 format: gl.RGBA,
-                attachment: this._exportTexture
+                attachment: this._texture
             }
         ];
         if (this._framebuffer) {
             twgl.resizeFramebufferInfo(gl, this._framebuffer, attachments, width, height);
-            twgl.resizeFramebufferInfo(gl, this._silhouetteBuffer, [{format: gl.RGBA}], width, height);
         } else {
             this._framebuffer = twgl.createFramebufferInfo(gl, attachments, width, height);
-            this._silhouetteBuffer = twgl.createFramebufferInfo(gl, [{format: gl.RGBA}], width, height);
         }
 
         gl.clearColor(0, 0, 0, 0);
         gl.clear(gl.COLOR_BUFFER_BIT);
 
         this._silhouettePixels = new Uint8Array(Math.floor(width * height * 4));
-        this._silhouetteImageData = this._canvas.getContext('2d').createImageData(width, height);
+        this._silhouetteImageData = new ImageData(width, height);
 
         this._silhouetteDirty = true;
     }
 
-    /**
-     * Set context state to match provided pen attributes.
-     * @param {CanvasRenderingContext2D} context - the canvas rendering context to be modified.
-     * @param {PenAttributes} penAttributes - the pen attributes to be used.
-     * @private
-     */
-    _setAttributes (context, penAttributes) {
-        penAttributes = penAttributes || DefaultPenAttributes;
-        const color4f = penAttributes.color4f || DefaultPenAttributes.color4f;
-        const diameter = penAttributes.diameter || DefaultPenAttributes.diameter;
-
-        const r = Math.round(color4f[0] * 255);
-        const g = Math.round(color4f[1] * 255);
-        const b = Math.round(color4f[2] * 255);
-        const a = color4f[3]; // Alpha is 0 to 1 (not 0 to 255 like r,g,b)
-
-        context.strokeStyle = `rgba(${r},${g},${b},${a})`;
-        context.lineCap = 'round';
-        context.lineWidth = diameter;
-    }
-
     /**
      * If there have been pen operations that have dirtied the canvas, update
      * now before someone wants to use our silhouette.
      */
     updateSilhouette () {
         if (this._silhouetteDirty) {
-            if (this._canvasDirty) {
-                this._drawToBuffer();
-            }
-
-            // Render export texture to another framebuffer
-            const gl = this._renderer.gl;
-
-            this._renderer.enterDrawRegion(this._toBufferDrawRegionId);
```

Comment:
We want to make sure we are always using draw regions for consistency/protection, so rather than changing this to be like clear (by moving bindFrameBufferInfo outside of the enterDrawRegion function), we'd rather change clear to call enterDrawRegion.

Can you explain the logic behind removing the `this._drawRectangleRegionEnter(this._stampShader, this._bounds);` step, which used t be called by enterDrawRegion? Are those always guaranteed to be no-ops? What makes sure that state is set?

Reply:
It looks like `_drawRectangleRegionEnter` sets the viewport and shader.

Calling `clear` on the `PenSkin` just calls `gl.clear`, which requires the framebuffer to be set but does *not* depend on the current viewport or shader.

Likewise for `updateSilhouette` which calls `gl.readPixels`, which also only requires the correct framebuffer to be bound.

The code for stamping lives in `RenderWebGL` and never called this code so it shouldn't be affected--it uses the same code path as drawing to the stage and sets the viewport + framebuffer itself.

Finally, the code for pen lines uses an existing draw region which sets the viewport, framebuffer, and shader.